Add conversion for rol and ror operators to c output

TGWDB opened the pull request:
This adds conversion for the rol and ror operators from CBMC
internals into C code. This is done by bit-twiddling the values
to achieve the outcome.

Fixes #6241

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

TGWDB committed:
Add conversion for rol and ror operators to c output

This adds conversion for the rol and ror operators from CBMC
internals into C code. This is done by bit-twiddling the values
to achieve the outcome.

Fixes #6241

TGWDB committed:
Add regression tests for rol/ror

This adds regression tests for rol/ror using symtba2gb and
goto-instrument.

TGWDB committed:
Add unit tests for expr2c for rol/ror

Adds unit test for expr2c of rol and ror operators, testing both
signed and unisgned conversions.

martin-cs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3794,6 +3794,49 @@ std::string expr2ct::convert_with_precedence(
 
+  else if(src.id() == ID_rol)
+  {
+    // AAAA rol n == (AAAA << n) | (AAAA >> (width(AAAA) - n))
```
What if `n` is greater than bit width?

martin-cs replied:
Also note the cases when shift can give undefined behaviour.

TGWDB replied:
This is done with modulo now.

thomasspriggs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3794,6 +3794,77 @@ std::string expr2ct::convert_with_precedence(
 
+  else if(src.id() == ID_rol)
+  {
+    // AAAA rol n ==   (AAAA << n % width(AAAA))
```
:no_entry_sign: The majority of the other conversions in `convert_with_precedence` are either one liners, or separated out into other functions. This new code should be factored out in the same way. The new function will need to be a member function of the `expr2ct` class (see `src/ansi-c/expr2c_class.h`) so that it can include the recursive call to `convert_with_precedence`. The `convert_with_precedence` function is a bit long as it is, lets not add to the problem.

thomasspriggs replied:
I now consider this to be sorted.

thomasspriggs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3794,6 +3794,77 @@ std::string expr2ct::convert_with_precedence(
+  }
+  else if(src.id() == ID_ror)
+  {
+    // AAAA rol n ==   (AAAA >> n % width(AAAA))
```
:no_entry_sign: This duplicated code will make maintenance more difficult. I suggest deduplicating it by parameterising the new function I suggested in my previous comment.
:bulb: I suggest creating a single use `left_or_rightt` enum class with left and right values to use as the type of the `left_or_right` parameter of the new function. This could alternatively be done without using a new enum with something like `bool is_right`, with true = right and false = left. But I think the enum would make for more self explanatory code.

thomasspriggs replied:
I now consider this to be sorted.

thomasspriggs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3794,6 +3794,77 @@ std::string expr2ct::convert_with_precedence(
+    {
+      // Shifts in C are arithmetic and care about sign, by forcing
+      // a cast to unsigned we force the shifts to perform rol/r
+      size_t ws = op0.type().get_size_t("width");
```
:pick: This and all other variables in this new code which are not mutated should be declared `const`.

thomasspriggs replied:
:no_entry_sign: Use `bitvector_typet::get_width()` instead of `get_size_t("width")`. By using `get_size_t` you are directly accessing the underlying data structure of the type. Fixing the accessibility to underlying `irept` would be a non-trivial refactor, but we should still use the well defined external interface where possible.

thomasspriggs replied:
:pick: What is `ws` supposed to be short for? Reminder - the coding standards state that we should "Avoid abbreviations".

thomasspriggs replied:
This looks to be sorted, with the exception of the unreachable code.

thomasspriggs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3794,6 +3794,77 @@ std::string expr2ct::convert_with_precedence(
+  {
+    // AAAA rol n ==   (AAAA << n % width(AAAA))
+    //               | (AAAA >> (width(AAAA) - n % width(AAAA)))
+    auto bin_expr = to_binary_expr(src);
```
:pick: I suggest renaming `bin_expr` to `rotate_expr`, because at this point `bin_expr` can only be a rotate expression, even if the type only specifies that it is a `binary_exprt`.

thomasspriggs replied:
:pick: I suggest swapping out `to_binary_expr(src)` with `expr_checked_cast<shift_exprt>(src)`. That way we are using a more specific type of expression and we are using the checks which are defined alongside that expression.

thomasspriggs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3794,6 +3794,77 @@ std::string expr2ct::convert_with_precedence(
+    }
+    // Get n because we need its type to construct the "width(AAAA)"
+    // value to use in operations below.
+    exprt &op1 = bin_expr.op1();
```
:pick: I suggest renaming `op1` to `distance`.

thomasspriggs replied:
:pick: Use `.distance()` of `shift_exprt` instead of `.op1()`, because this is the specifically named getter for this data member.

thomasspriggs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3794,6 +3794,77 @@ std::string expr2ct::convert_with_precedence(
+    auto bin_expr = to_binary_expr(src);
+    // Get AAAA and if it's signed wrap it in a cast to remove
+    // the sign since this messes with C shifts
+    exprt &op0 = bin_expr.op0();
```
:pick: Use `.op()` of `shift_exprt` rather than `op0()`, because this is the specifically named getter for this data member.

thomasspriggs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3794,6 +3794,77 @@ std::string expr2ct::convert_with_precedence(
+    exprt &op1 = bin_expr.op1();
+    // Construct the "width(AAAA)" constant
+    irep_idt width_val = op0.type().get("width");
+    auto width = constant_exprt(width_val, op1.type());
```
:no_entry_sign: `from_integer` should be used to produce the `constant_exprt`, rather than using the constructor of `constant_exprt` directly. This is because the string based representation is not guaranteed to be stable. I believe we have made changes such as changing the base used for this representation before now. `from_integer` can take the `size_t` extracted via the `get_width` call, along with the same `typet`, which you are currently using.

thomasspriggs replied:
:pick: I suggest renaming this `width` to `width_expression`, in order to better differentiate it from the value of the expression.

thomasspriggs replied:
This looks to be sorted.

thomasspriggs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3794,6 +3794,77 @@ std::string expr2ct::convert_with_precedence(
+    // value to use in operations below.
+    exprt &op1 = bin_expr.op1();
+    // Construct the "width(AAAA)" constant
+    irep_idt width_val = op0.type().get("width");
```
:pick: Duplicated extraction of getting the width from `op0`. It should be possible to move the previous getting of the width out of the `if` statement and reuse it. We can expect the type to be a `bitvector_typet` whether or not it is signed. So extracting the width can be - `const size_t width = type_checked_cast<bitvector_typet>(op0.type()).get_width();`

thomasspriggs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -3794,6 +3794,77 @@ std::string expr2ct::convert_with_precedence(
+    auto width = constant_exprt(width_val, op1.type());
+    // Apply modulo to n since shifting will overflow
+    // That is: 0001 << 4 == 0, but 0001 rol 4 == 0001
+    op1 = mod_exprt(op1, width);
```
:pick: I suggest creating a new variable named `distance_modulo_width` instead, in order to better specify what this is and in order to avoid mutating.

thomasspriggs reviewed (changes requested):
It is good that we have a fix for the issue and that we have both regression and unit tests. However I think there are some issues which should be addressed for the sake of maintainability. The issues which I think really need to be addressed before we merge are marked with :no_entry_sign: Hopefully having the tests should make it easy to check the refactorings are sound as you address the issues.

TGWDB committed:
Tidy up and address comments

Tidies up the style and addresses reviewer comments:
- use to_shift_exprt before call to convert_rox
- remove "kind" path to try and guess bit width

TGWDB committed:
Clang format

chrisr-diffblue commented on src/ansi-c/expr2c.cpp:
```diff
@@ -2290,6 +2290,71 @@ std::string expr2ct::convert_initializer_list(const exprt &src)
+    mod_exprt(rotate_expr.distance(), width_expr);
+  // Now put the pieces together
+  // width(AAAA) - (n % width(AAAA))
+  const auto complement_expr = minus_exprt(width_expr, distance_modulo_width);
```
I don't think this is actually the _complement expression_, is the complement _width_ expression ?

TGWDB replied:
Fixed

chrisr-diffblue commented on unit/ansi-c/expr2c.cpp:
```diff
@@ -0,0 +1,66 @@
+
+TEST_CASE("rol_2_c_conversion_unsigned", "[core][ansi-c][expr2c]")
+{
+  auto lhs = from_integer(31, unsignedbv_typet(32));
```
Personal preference here :-) But for tests used for bit-twiddling operations, I'd be rather tempted to use hex representations of the various immediate constants in tests like this. Might just be my background, but I find that easier to mentally model, rather than also adding a decimal->hex->binary mental translation :-)

chrisr-diffblue commented on unit/ansi-c/expr2c.cpp:
```diff
@@ -0,0 +1,66 @@
+  "[core][ansi-c][expr2c][convert_with_precedence]")
+{
+  config.ansi_c.mode = configt::ansi_ct::flavourt::GCC;
+  config.ansi_c.set_arch_spec_i386();
```
Might be worth a comment explaining why this is necessary?

TGWDB replied:
Done

chrisr-diffblue reviewed (approved):
Mostly looks OK, though I think @thomasspriggs has some blocking review comments that it would be good to resolve. Also, how cross-platform/cross-architecture/cross endian is this work? In particular, are we confident this works on, say ARM architecture where its switchable big/little endian? I _think_ cbmc has `--little-endian` and `--big-endian` options.

tautschnig commented on src/ansi-c/expr2c.cpp:
```diff
@@ -2290,6 +2290,72 @@ std::string expr2ct::convert_initializer_list(const exprt &src)
+  //               | (AAAA rhs_op (width(AAAA) - n % width(AAAA)))
+  // Where lhs_op and rhs_op depend on whether it is rol or ror
+  // auto rotate_expr = to_binary_expr(src);
+  auto rotate_expr = expr_checked_cast<shift_exprt>(src);
```
In keeping with other functions I'd prefer for `convert_rox` to take a `shift_exprt` as first argument and `to_shift_expr` being used at the call site.

tautschnig commented on src/ansi-c/expr2c.cpp:
```diff
@@ -2290,6 +2290,72 @@ std::string expr2ct::convert_initializer_list(const exprt &src)
+  }
+  else
+  {
+    // Type that can't be represented as bitvector
+    // get some kind of width in a potentially unsafe way
+    type_width = op0.type().get_size_t("width");
```
I think you should just give up in that `else` case. You might well get back a `nil` irep, which will make the `from_integer` fail. You're trying to be very kind, but I don't think you're doing the user any favour.

thomasspriggs replied:
Agreed. I'd just put an `UNREACHABLE` here. I think the comments from codecov suggest that it is in fact unreachable with the way it is currently used.

thomasspriggs reviewed (approved):
My blocking comments have been addressed. Therefore I am going to approve this PR.

thomasspriggs commented on src/ansi-c/expr2c.cpp:
```diff
@@ -2329,23 +2329,24 @@ std::string expr2ct::convert_rox(const exprt &src, unsigned precedence)
-    // AAAA << complement_expr
-    rhs_expr = shl_exprt(op0, complement_expr);
+    // AAAA << complement_width_expr
+    rhs_expr = shl_exprt(op0, complement_width_expr);
```
:pick: These changes appear to have been squashed into the wrong commit.

TGWDB replied:
Fixed.
